Parser generator fails with InvalidOperationException: Incompatible fold functions

**Describe the bug**
When alternatives for left-recursive rules are present and they don't match the same parsing type, they need to be attached to different transformer functions. However, this causes the parser generator to emit an InvalidOperatorException with message "Incompatible fold functions", making this kind of rule non-representable in the parser.

Consider the following rule:
```
call : primary ( '(' args ')' | '.' Identifier )*
```
Because the alternatives `'(' args ')'` and `'.' Identifier` have different types (notice the different arity), they need to use two different  left-recursive transformator functions:
```cs
[Rule("call : call '(' (expression (',' expression)*)? ')'")]
public static Expression FunctionCall(Expression callee, Token _open, Punctuated<Expression, Token> arguments, Token _close) =>
            new Expression.Call(callee, arguments.Values.ToList().AsValue());

[Rule("call : call '.' Identifier")]
public static Expression MemberAccess(Expression @object, Token _dot, Token identifier) =>
            new Expression.MemberGet(@object, identifier.Text);

[Rule("call : primary")]
public static Expression Identity(Expression expression) => expression;
```

However, this structure results in the following compile-time error:
```
warning CS8785: Generator 'ParserSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'InvalidOperationException' with message 'Incompatible fold functions
```

**Which libraries does it affect?**
`Yoakke.Parser.Generator` version `2021.8.5-2.33.35-nightly`.

**To Reproduce**
1. Make a `Parser` with the rules and transformator functions shown above.
2. Make sure the parser generator is installed and working.
3. Build the project.
4. Observe the build fail as the parser fails to generate.

**Expected behavior**
The parser generates normally and is capable of parsing expressions like `x.y.z().g()().w` as specified by the `call` rules.

**Environment (please complete the following information):**
 - OS: Windows 10.0.19043
 - .NET version: 5.0.8 (35964c9215), SDK 5.0.302
 - Yoakke.Parser.Generator version: 2021.8.5-2.33.35-nightly

--- a/Sources/Core/Yoakke.Parser.Generator/Ast/BnfAst.FoldLeft.cs
+++ b/Sources/Core/Yoakke.Parser.Generator/Ast/BnfAst.FoldLeft.cs
@@ -3,6 +3,8 @@
 // Source repository: https://github.com/LanguageDev/Yoakke
 
 using System;
+using System.Collections.Generic;
+using System.Linq;
 using Microsoft.CodeAnalysis;
 using Yoakke.Utilities.Compatibility;
 
@@ -22,46 +24,45 @@ namespace Yoakke.Parser.Generator.Ast
             public BnfAst First { get; }
 
             /// <summary>
-            /// The element to apply repeatedly after.
+            /// The alternative elements to apply repeatedly after.
             /// </summary>
-            public BnfAst Second { get; }
-
-            /// <summary>
-            /// The transformation method that does the folding.
-            /// </summary>
-            public IMethodSymbol Method { get; }
+            public IReadOnlyList<(BnfAst Node, IMethodSymbol Method)> Seconds { get; }
 
             /// <summary>
             /// Initializes a new instance of the <see cref="FoldLeft"/> class.
             /// </summary>
             /// <param name="first">The first element of the fold.</param>
-            /// <param name="second">The second element of the fold, that will be repeated.</param>
-            /// <param name="method">The folding transformation.</param>
-            public FoldLeft(BnfAst first, BnfAst second, IMethodSymbol method)
+            /// <param name="seconds">The second elements and transformations of the fold, that will be repeated.</param>
+            public FoldLeft(BnfAst first, IReadOnlyList<(BnfAst Node, IMethodSymbol Method)> seconds)
             {
                 this.First = first;
-                this.Second = second;
-                this.Method = method;
+                this.Seconds = seconds;
             }
 
+            // TODO: Probably not corret
             /// <inheritdoc/>
             public override bool Equals(BnfAst other) => other is FoldLeft fl
                 && this.First.Equals(fl.First)
-                && this.Second.Equals(fl.Second)
-                && SymbolEqualityComparer.Default.Equals(this.Method, fl.Method);
+                && this.Seconds.SequenceEqual(fl.Seconds);
+
+            // TODO: Probably not corret
+            /// <inheritdoc/>
+            public override int GetHashCode() => this.First.GetHashCode();
 
             /// <inheritdoc/>
-            public override int GetHashCode() => HashCode.Combine(this.First, this.Second, this.Method);
-
-            /// <inheritdoc/>
-            public override BnfAst Desugar() => new FoldLeft(this.First.Desugar(), this.Second.Desugar(), this.Method);
+            public override BnfAst Desugar() => new FoldLeft(
+                this.First.Desugar(),
+                this.Seconds.Select(s => (s.Node.Desugar(), s.Method)).ToList());
 
             /// <inheritdoc/>
             public override string GetParsedType(RuleSet ruleSet, TokenKindSet tokens)
             {
                 var firstType = this.First.GetParsedType(ruleSet, tokens);
-                var mappedType = this.Method.ReturnType.ToDisplayString();
-                if (firstType != mappedType) throw new InvalidOperationException("Incompatible folded types");
+                foreach (var (_, method) in this.Seconds)
+                {
+                    var mappedType = method.ReturnType.ToDisplayString();
+                    if (firstType != mappedType) throw new InvalidOperationException("Incompatible folded types");
+                }
                 return firstType;
             }
         }

--- a/Sources/Core/Yoakke.Parser.Generator/BnfDesugar.cs
+++ b/Sources/Core/Yoakke.Parser.Generator/BnfDesugar.cs
@@ -89,9 +89,8 @@ namespace Yoakke.Parser.Generator
 
         private static BnfAst EliminateLeftRecursionInAlternation(Rule rule, BnfAst.Alt alt)
         {
-            var alphas = new List<BnfAst>();
+            var alphas = new List<(BnfAst Node, IMethodSymbol Method)>();
             var betas = new List<BnfAst>();
-            IMethodSymbol? fold = null;
             foreach (var child in alt.Elements)
             {
                 // If the inside has no transformation, we don't care
@@ -103,15 +102,7 @@ namespace Yoakke.Parser.Generator
                 // We found a left-recursive sequence inside an alternative, add it as alpha
                 if (transform.Subexpr is BnfAst.Seq seq && TrySplitLeftRecursion(rule, seq, out var alpha))
                 {
-                    if (fold == null)
-                    {
-                        fold = transform.Method;
-                    }
-                    else if (!SymbolEqualityComparer.Default.Equals(fold, transform.Method))
-                    {
-                        throw new InvalidOperationException("Incompatible fold functions");
-                    }
-                    alphas.Add(alpha!);
+                    alphas.Add((alpha!, transform.Method));
                 }
                 else
                 {
@@ -121,8 +112,7 @@ namespace Yoakke.Parser.Generator
             if (alphas.Count == 0 || betas.Count == 0) return alt;
             // We have left-recursion
             var betaNode = betas.Count == 1 ? betas[0] : new BnfAst.Alt(betas);
-            var alphaNode = alphas.Count == 1 ? alphas[0] : new BnfAst.Alt(alphas);
-            return new BnfAst.FoldLeft(betaNode, alphaNode, fold!);
+            return new BnfAst.FoldLeft(betaNode, alphas);
         }
 
         private static bool TrySplitLeftRecursion(Rule rule, BnfAst.Seq seq, [MaybeNullWhen(false)] out BnfAst? alpha)

--- a/Sources/Core/Yoakke.Parser.Generator/ParserSourceGenerator.cs
+++ b/Sources/Core/Yoakke.Parser.Generator/ParserSourceGenerator.cs
@@ -192,7 +192,7 @@ namespace {namespaceName}
         {
             BnfAst.Alt alt => alt.Elements.All(this.CheckBnfAst),
             BnfAst.Seq seq => seq.Elements.All(this.CheckBnfAst),
-            BnfAst.FoldLeft foldl => this.CheckBnfAst(foldl.First) && this.CheckBnfAst(foldl.Second),
+            BnfAst.FoldLeft foldl => this.CheckBnfAst(foldl.First) && foldl.Seconds.Select(s => s.Node).All(this.CheckBnfAst),
             BnfAst.Opt opt => this.CheckBnfAst(opt.Subexpr),
             BnfAst.Group grp => this.CheckBnfAst(grp.Subexpr),
             BnfAst.Rep0 rep0 => this.CheckBnfAst(rep0.Subexpr),
@@ -253,21 +253,36 @@ namespace {namespaceName}
 
             case BnfAst.FoldLeft fold:
             {
-                var binder = this.GetTopLevelPattern(fold.Second);
-                var flattenedValues = FlattenBind(binder);
                 var firstVar = this.GenerateBnf(code, rule, fold.First, lastIndex);
-                var call = $"{fold.Method.Name}({resultVar}.Ok.Value, {flattenedValues})";
+                var nextResultVar = this.AllocateVarName();
                 code.AppendLine($"if ({firstVar}.IsOk) {{");
                 code.AppendLine($"    {resultVar} = {firstVar};");
                 code.AppendLine("    while (true) {");
-                var secondVar = this.GenerateBnf(code, rule, fold.Second, $"{resultVar}.Ok.Offset");
-                var errMerge = $"{TypeNames.ParseError}.Unify({resultVar}.FurthestError, {secondVar}.FurthestError)";
-                code.AppendLine($"        if ({secondVar}.IsError) {{");
-                code.AppendLine($"            {resultVar} = {TypeNames.ParserBase}.MergeError({resultVar}.Ok, {secondVar}.Error);");
+                code.AppendLine($"        var {nextResultVar} = {resultVar};");
+
+                foreach (var (second, method) in fold.Seconds)
+                {
+                    var binder = this.GetTopLevelPattern(second);
+                    var flattenedValues = FlattenBind(binder);
+                    var call = $"{method.Name}({resultVar}.Ok.Value, {flattenedValues})";
+                    var secondVar = this.GenerateBnf(code, rule, second, $"{resultVar}.Ok.Offset");
+                    var errMerge = $"{TypeNames.ParseError}.Unify({nextResultVar}.FurthestError, {secondVar}.FurthestError)";
+                    var makeOk = $"{TypeNames.ParserBase}.Ok({call}, {secondVar}.Ok.Offset, {errMerge})";
+
+                    code.AppendLine($"if ({secondVar}.IsError) {{");
+                    code.AppendLine($"    {nextResultVar} = {TypeNames.ParserBase}.MergeError({nextResultVar}.Ok, {secondVar}.Error);");
+                    code.AppendLine("} else {");
+                    code.AppendLine($"    var {binder} = {secondVar}.Ok.Value;");
+                    code.AppendLine($"    {nextResultVar} = {TypeNames.ParserBase}.MergeAlternatives({makeOk}, {nextResultVar});");
+                    code.AppendLine("}");
+                }
+
+                code.AppendLine($"        if ({nextResultVar}.Ok.Offset == {resultVar}.Ok.Offset) {{");
+                code.AppendLine($"            {resultVar} = {nextResultVar};");
                 code.AppendLine("            break;");
+                code.AppendLine("        } else {");
+                code.AppendLine($"            {resultVar} = {nextResultVar};");
                 code.AppendLine("        }");
-                code.AppendLine($"        var {binder} = {secondVar}.Ok.Value;");
-                code.AppendLine($"        {resultVar} = {TypeNames.ParserBase}.Ok({call}, {secondVar}.Ok.Offset, {errMerge});");
                 code.AppendLine("    }");
                 code.AppendLine("} else {");
                 code.AppendLine($"    {resultVar} = {firstVar}.Error;");

--- a/Sources/Core/Yoakke.Parser/ParserBase.cs
+++ b/Sources/Core/Yoakke.Parser/ParserBase.cs
@@ -34,6 +34,9 @@ namespace Yoakke.Parser
         /// <param name="tokens">The tokens to lex.</param>
         protected ParserBase(IEnumerable<IToken> tokens)
         {
+            // TODO: We could eliminate the null lexer by implementing a lexer
+            // that simply feeds off of an IEnumerable like this
+            // Whould get rid of a nasty null-check
             // TODO: Make this lazy instead of loading it into the peek-buffer?
             this.peek = new RingBuffer<IToken>();
             foreach (var token in tokens) this.peek.AddBack(token);

--- a/Sources/Tests/Yoakke.Parser.Tests/LeftRecursionAltsIssueTests.cs
+++ b/Sources/Tests/Yoakke.Parser.Tests/LeftRecursionAltsIssueTests.cs
@@ -1,0 +1,68 @@
+// Copyright (c) 2021 Yoakke.
+// Licensed under the Apache License, Version 2.0.
+// Source repository: https://github.com/LanguageDev/Yoakke
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Yoakke.Lexer;
+using Yoakke.Lexer.Attributes;
+using Yoakke.Parser.Attributes;
+using IgnoreAttribute = Yoakke.Lexer.Attributes.IgnoreAttribute;
+using Token = Yoakke.Lexer.IToken<Yoakke.Parser.Tests.LeftRecursionAltsIssueTests.TokenType>;
+
+namespace Yoakke.Parser.Tests
+{
+    [Parser(typeof(LeftRecursionAltsIssueTests.TokenType))]
+    internal partial class LeftRecursionAltsIssueParser
+    {
+        [Rule("expression : call")]
+        [Rule("call : primary")]
+        public static string Identity(string expression) => expression;
+
+        [Rule("call : call '(' ')'")]
+        public static string FunctionCall(string callee, Token _open, Token _close) => $"({callee}())";
+
+        [Rule("call : call '.' Identifier")]
+        public static string MemberAccess(string @object, Token _dot, Token identifier) => $"({@object}.{identifier.Text})";
+
+        [Rule("primary : Identifier")]
+        public static string Primary(Token atom) => atom.Text;
+    }
+
+    [TestClass]
+    public class LeftRecursionAltsIssueTests
+    {
+        [Lexer("LeftRecursionAltsIssueLexer")]
+        public enum TokenType
+        {
+            [End] End,
+            [Error] Error,
+            [Ignore] [Regex(Regexes.Whitespace)] Whitespace,
+
+            [Regex(Regexes.Identifier)] Identifier,
+
+            [Token(".")] Dot,
+            [Token("(")] OpenParen,
+            [Token(")")] CloseParen,
+        }
+
+        private static string Parse(string source) =>
+           new LeftRecursionAltsIssueParser(new LeftRecursionAltsIssueLexer(source)).ParseCall().Ok.Value;
+
+        [TestMethod]
+        public void TestPrimary() => Assert.AreEqual("x", Parse("x"));
+
+        [TestMethod]
+        public void TestCall() => Assert.AreEqual("(x())", Parse("x()"));
+
+        [TestMethod]
+        public void TestMemberAccess() => Assert.AreEqual("(x.y)", Parse("x.y"));
+
+        [TestMethod]
+        public void TestLongChain() => Assert.AreEqual("(((((((x.y)())()).z).w)()).q)", Parse("x.y()().z.w().q"));
+    }
+}

